src/Rings/affine-algebra-homs.jl

```diff
@@ -0,0 +1,264 @@
+   return map_poly(f, p)
+end
+
+###############################################################################
+#
+#   Preimage and Kernel for Identity Algebra Homomorphisms
+#
+###############################################################################
+
+function preimage(f::Map(IdAlgHom), I::U) where U <: Union{MPolyIdeal, MPolyQuoIdeal}
+   @assert base_ring(I) == domain(f)
+   return I
+end
+
+function kernel(f::Map(IdAlgHom))
+   return f.kernel
+end
+
+###############################################################################
+#
+#   AlgebraHomomorphism
+#
+###############################################################################
+
+# this terrible function is used to get polynomials in and out of Singular and
+# to change: polys over the quotient ring <=> polys over the non-quotient ring
+function _badpolymap(f, R::MPolyRing)
+  parent(f) == R && return f
+  @assert ngens(parent(f)) == ngens(R)
+  B = base_ring(R)
+  g = MPolyBuildCtx(R)
+  for (c, e) = zip(Nemo.coeffs(f), Nemo.exponent_vectors(f))
+    push_term!(g, B(c), e)
+  end
+  return finish(g)
+end
+
+mutable struct AlgHom{T} <: AbstractAlgebra.Map{Ring, Ring,
+         AbstractAlgebra.SetMap, AlgHom} where T <: Union{AbstractAlgebra.Ring, AbstractAlgebra.Field}
+   domain::Union{MPolyRing, MPolyQuo}
+   codomain::Union{MPolyRing, MPolyQuo}
+   image::Vector{U} where U <: Union{MPolyElem, MPolyQuoElem}
+   salghom::Singular.SAlgHom
+   kernel::Union{MPolyIdeal, MPolyQuoIdeal}
+
+   function AlgHom{T}(R::U, S::W, V::Vector{X}) where {T, U, W, X}
+      Rx = singular_ring(R)
+      Sx = singular_ring(S)
+      if isdefined(S, :I) ## Check if S is a quotient ring
+         Vx = map(p -> _badpolymap(S(p).f, Sx), V)
+      else
+         Vx = map(p -> _badpolymap(S(p), Sx), V)
+      end
+
+      z = new(R, S, V, Singular.AlgebraHomomorphism(Rx, Sx, Vx))
+      return z
+   end
+
+   function AlgHom{T}(R::U, S::W, V::Vector{X}, phi::Singular.SAlgHom) where {T, U, W, X}
+      return new(R, S, V, phi)
+   end
+end
+
+###############################################################################
+#
+#   I/O for Algebra Homomorphisms
+#
+###############################################################################
+
+function show(io::IO, M::Map(AlgHom))
+   println(io, "Algebra homomorphism with")
+   println(io, "")
+   println(io, "domain: ", domain(M))
+   println(io, "")
+   println(io, "codomain: ", codomain(M))
+   println(io, "")
+   println(io, "defining equations: ", M.image)
+end
+
+###############################################################################
+#
+#   Algebra Homomorphism constructor
+#
+###############################################################################
+
+#Compute the type of the base_ring of the underlying poly ring
+function _type_helper(R)
+   if isdefined(R, :I)
+      return typeof(base_ring(R.R))
+   else
+      return typeof(base_ring(R))
+   end
+end
```

Comment:
I don't think this is a good idea.

Reply:
What do you mean? 
I just need to know over which field we consider our ring as an algebra and base_ring returns the underlying polynomial ring if it is a MPolyQuo.

Reply:
I am also confused what this does: you check for the presence of a field `R.I` and if it is there, access a different field `R.R`? And which type(s) is `R` even allowed to have? Looking at the call in `AlgebraHomomorphism`, it seems `D` is either an `MPolyRing{T}` or a `MPolyQuo`. In which of those is `R.I` present? My guess/hope is it is present exactly in `MPolyQuo`? If so, then far better would to dispatch on that type... 

In any case, if it works right now, that's OK, and we can "improve" it later on